clusterloader2: Bump Kubernetes dependency to v1.28.12

mboersma opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This updates the old vendoring of k8s.io in clusterloader2 to a supported Kubernetes version.

#2795 was already open, and I wanted to investigate the remaining issues. I ended up going minor-by-minor release to figure out (mostly) what changes to make, although I still have a `// TODO` in there.

#### Which issue(s) this PR fixes:

Refs #2795

#### Special notes for your reviewer:

@wojtek-t please copy any relevant changes from this PR to yours! I'll be happy to review it.

mboersma commented:
/hold
/retitle [WIP] clusterloader2: Bump Kubernetes dependency to v1.28.12

I don't intend to merge this, just want to see how CI treats it and help #2795 get over the finish line.

mboersma commented on clusterloader2/pkg/framework/config/client_config.go:
```diff
@@ -107,7 +107,7 @@ func transportHack(config *restclient.Config) error {
 	})
 	config.WrapTransport = transportConfig.WrapTransport
-	config.Dial = transportConfig.Dial
+	// config.Dial = transportConfig.Dial // TODO: What replaces this?
```
I wasn't sure how to handle this new struct mismatch.

wojtek-t replied:
https://github.com/kubernetes/kubernetes/commit/3313a70d5bcc40a39f99f482c18effc9de6072ba shows what to do here

wojtek-t reviewed (commented):
Thanks a lot for opening this - let's proceed with your PR - I will review it and close mine.

wojtek-t commented on clusterloader2/pkg/measurement/common/metrics/metrics_grabber.go:
```diff
@@ -195,7 +195,7 @@ func (g *Grabber) GrabFromControllerManager() (ControllerManagerMetrics, error)
 	podName := fmt.Sprintf("%v-%v", "kube-controller-manager", g.masterName)
 	g.waitForControllerManagerReadyOnce.Do(func() {
-		if readyErr := e2epod.WaitTimeoutForPodReadyInNamespace(g.client, podName, metav1.NamespaceSystem, framework.PodStartTimeout); readyErr != nil {
+		if readyErr := e2epod.WaitTimeoutForPodReadyInNamespace(context.TODO(), g.client, podName, metav1.NamespaceSystem, framework.PodStartTimeout); readyErr != nil {
```
nit: context.Background()

wojtek-t commented on clusterloader2/pkg/measurement/util/informer/informer.go:
```diff
@@ -61,7 +61,7 @@ func NewDynamicInformer(
 	handleObj func(interface{}, interface{}),
 ) {
-	i.AddEventHandler(cache.ResourceEventHandlerFuncs{
+	_, _ = i.AddEventHandler(cache.ResourceEventHandlerFuncs{
```
let's not ignore error - let's at least log an error in such case

wojtek-t commented on clusterloader2/pkg/provider/autopilot.go:
```diff
@@ -60,7 +62,7 @@ func (p *AutopilotProvider) GetConfig() Config {
 func (p *AutopilotProvider) RunSSHCommand(cmd, host string) (string, string, int, error) {
 	// gce provider takes ssh key from GCE_SSH_KEY.
-	r, err := sshutil.SSH(cmd, host, "gce")
+	r, err := sshutil.SSH(context.TODO(), cmd, host, "gce")
```
context.Background()

same for other files below

wojtek-t reviewed (commented):
This is great - I just added couple minor comments

wojtek-t commented:
/hold cancel

mboersma commented:
/retitle clusterloader2: Bump Kubernetes dependency to v1.28.12

I updated based on review comments and squashed the commits. Let me know if there's anything else to change.

wojtek-t commented:
/lgtm
/approve
